### PR TITLE
Acknowledge version release types from mod platforms

### PR DESCRIPTION
**Pull request by Ryex**
- add a `verison_type` aka release type to `IndexedVersion`
-  display version type during update and download selection
- when sorting `IndexVersion`  give presadence to better release types

image

image

### CODE REVIEW

**Comment by getchoo** [comment]
i'm going to assume this is due to needing a rebase, but currently if i try to update mods from modrinth i get the following log

```
    11.199 D | [launcher.task.net.upload]: "{4a4e9946-8453-42e3-94c0-7639c3f461f3}" Uploading  "https://api.modrinth.com/v2/version_files"
    11.455 C | [launcher.task.net.upload]: "{4a4e9946-8453-42e3-94c0-7639c3f461f3}" Failed  "https://api.modrinth.com/v2/version_files"  with reason  QNetworkReply::ProtocolInvalidOperationError
    11.456 D | [launcher.task.net.upload]: "{4a4e9946-8453-42e3-94c0-7639c3f461f3}" Upload failed in previous step: "https://api.modrinth.com/v2/version_files"
    11.456 D | [launcher.task.net.upload]: "{4a4e9946-8453-42e3-94c0-7639c3f461f3}" Uploading  "https://api.modrinth.com/v2/version_files"
    11.594 C | [launcher.task.net.upload]: "{4a4e9946-8453-42e3-94c0-7639c3f461f3}" Failed  "https://api.modrinth.com/v2/version_files"  with reason  QNetworkReply::ProtocolInvalidOperationError
    11.594 D | [launcher.task.net.upload]: "{4a4e9946-8453-42e3-94c0-7639c3f461f3}" Upload failed in previous step: "https://api.modrinth.com/v2/version_files"
    11.594 D | [launcher.task.net.upload]: "{4a4e9946-8453-42e3-94c0-7639c3f461f3}" Uploading  "https://api.modrinth.com/v2/version_files"
    11.691 C | [launcher.task.net.upload]: "{4a4e9946-8453-42e3-94c0-7639c3f461f3}" Failed  "https://api.modrinth.com/v2/version_files"  with reason  QNetworkReply::ProtocolInvalidOperationError
    11.691 D | [launcher.task.net.upload]: "{4a4e9946-8453-42e3-94c0-7639c3f461f3}" Upload failed in previous step: "https://api.modrinth.com/v2/version_files"
    11.691 W | Error while parsing JSON response from Modrinth::CurrentVersions at  0  reason:  "illegal value"
    11.691 W | ""
    11.694 W | ""
    11.694 W | No addonId found!
```

curseforge doesn't seem to be affected though, and everything works as normal in the "Download Mods" menu. the only other thing i would recommend is changing the indicator to something like `<mod info> [version_type]` since `<mod info> : <version_type>` looks a little off to me

File: launcher/modplatform/ModIndex.h [hunk 1]
```diff
@@ -51,11 +52,46 @@ struct DonationData {
...
+
+    IndexedVersionType::Enum m_type;
+};
+
 struct IndexedVersion {
     QVariant addonId;
     QVariant fileId;
     QString version;
     QString version_number = {};
+    IndexedVersionType verison_type;
```
Comment (DioEgizio) [inline launcher/modplatform/ModIndex.h]: more like version_type

File: launcher/modplatform/CheckUpdateTask.h [hunk 2]
```diff
@@ -12,21 +12,39 @@ class CheckUpdateTask : public Task {
...
+                    std::optional<ResourceAPI::ModLoaderTypes> loaders,
+                    std::shared_ptr<ModFolderModel> mods_folder)
+        : Task(nullptr), m_mods(mods), m_game_versions(mcVersions), m_loaders(loaders), m_mods_folder(mods_folder){};
 
     struct UpdatableMod {
         QString name;
         QString old_hash;
         QString old_version;
         QString new_version;
+        std::optional<ModPlatform::IndexedVersionType> new_verison_type;
```
Comment (DioEgizio) [inline launcher/modplatform/CheckUpdateTask.h]: more like new_version_type

File: launcher/modplatform/CheckUpdateTask.h [hunk 3]
```diff
@@ -12,21 +12,39 @@ class CheckUpdateTask : public Task {
...
+                     QString new_v,
+                     std::optional<ModPlatform::IndexedVersionType> new_v_type,
+                     QString changelog,
+                     ModPlatform::ResourceProvider p,
+                     shared_qobject_ptr<ResourceDownloadTask> t)
+            : name(name)
+            , old_hash(old_h)
+            , old_version(old_v)
+            , new_version(new_v)
+            , new_verison_type(new_v_type)
```
Comment (DioEgizio) [inline launcher/modplatform/CheckUpdateTask.h]: more like new_version_type

File: launcher/modplatform/flame/FlameAPI.cpp [hunk 4]
```diff
@@ -144,7 +144,8 @@ auto FlameAPI::getLatestVersion(VersionSearchArgs&& args) -> ModPlatform::Indexe
             for (auto file : arr) {
                 auto file_obj = Json::requireObject(file);
                 auto file_tmp = FlameMod::loadIndexedPackVersion(file_obj);
-                if(file_tmp.date > ver_tmp.date) {
+                bool better_release = file_tmp.verison_type <= ver_tmp.verison_type;
```
Comment (DioEgizio) [inline launcher/modplatform/flame/FlameAPI.cpp]: ditto

File: launcher/modplatform/flame/FlameCheckUpdate.cpp [hunk 5]
```diff
@@ -173,7 +173,7 @@ void FlameCheckUpdate::executeTask()
             }
 
             auto download_task = makeShared<ResourceDownloadTask>(pack, latest_ver, m_mods_folder);
-            m_updatable.emplace_back(pack.name, mod->metadata()->hash, old_version, latest_ver.version,
+            m_updatable.emplace_back(pack.name, mod->metadata()->hash, old_version, latest_ver.version, latest_ver.verison_type,
```
Comment (DioEgizio) [inline launcher/modplatform/flame/FlameCheckUpdate.cpp]: ditto

File: launcher/modplatform/flame/FlameModIndex.cpp [hunk 6]
```diff
@@ -94,8 +94,9 @@ void FlameMod::loadIndexedPackVersions(ModPlatform::IndexedPack& pack,
     }
 
     auto orderSortPredicate = [](const ModPlatform::IndexedVersion& a, const ModPlatform::IndexedVersion& b) -> bool {
+        bool a_better_release = a.verison_type <= b.verison_type;
```
Comment (DioEgizio) [inline launcher/modplatform/flame/FlameModIndex.cpp]: ditto

File: launcher/modplatform/flame/FlameModIndex.cpp [hunk 7]
```diff
@@ -123,6 +124,7 @@ auto FlameMod::loadIndexedPackVersion(QJsonObject& obj, bool load_changelog) ->
     file.version = Json::requireString(obj, "displayName");
     file.downloadUrl = Json::ensureString(obj, "downloadUrl");
     file.fileName = Json::requireString(obj, "fileName");
+    file.verison_type = ModPlatform::IndexedVersionType(Json::requireInteger(obj, "releaseType"));
```
Comment (DioEgizio) [inline launcher/modplatform/flame/FlameModIndex.cpp]: ditto

File: launcher/modplatform/modrinth/ModrinthCheckUpdate.cpp [hunk 8]
```diff
@@ -161,8 +161,8 @@ void ModrinthCheckUpdate::executeTask()
 
                     auto download_task = makeShared<ResourceDownloadTask>(pack, project_ver, m_mods_folder);
 
-                    m_updatable.emplace_back(pack.name, hash, mod->version(), project_ver.version_number, project_ver.changelog,
-                                             ModPlatform::ResourceProvider::MODRINTH, download_task);
+                    m_updatable.emplace_back(pack.name, hash, mod->version(), project_ver.version_number, project_ver.verison_type,
```
Comment (DioEgizio) [inline launcher/modplatform/modrinth/ModrinthCheckUpdate.cpp]: ditto (also for the line after)

File: launcher/modplatform/modrinth/ModrinthPackIndex.cpp [hunk 9]
```diff
@@ -109,8 +109,9 @@ void Modrinth::loadIndexedPackVersions(ModPlatform::IndexedPack& pack,
             unsortedVersions.append(file);
     }
     auto orderSortPredicate = [](const ModPlatform::IndexedVersion& a, const ModPlatform::IndexedVersion& b) -> bool {
+        bool a_better_release = a.verison_type <= b.verison_type;
```
Comment (DioEgizio) [inline launcher/modplatform/modrinth/ModrinthPackIndex.cpp]: ditto

File: launcher/modplatform/modrinth/ModrinthPackIndex.cpp [hunk 10]
```diff
@@ -138,6 +139,8 @@ auto Modrinth::loadIndexedPackVersion(QJsonObject& obj, QString preferred_hash_t
     }
     file.version = Json::requireString(obj, "name");
     file.version_number = Json::requireString(obj, "version_number");
+    file.verison_type = ModPlatform::IndexedVersionType(Json::requireString(obj, "version_type"));
```
Comment (DioEgizio) [inline launcher/modplatform/modrinth/ModrinthPackIndex.cpp]: ditto

File: launcher/ui/dialogs/ModUpdateDialog.cpp [hunk 11]
```diff
@@ -362,6 +362,11 @@ void ModUpdateDialog::appendMod(CheckUpdateTask::UpdatableMod const& info)
     auto new_version_item = new QTreeWidgetItem(item_top);
     new_version_item->setText(0, tr("New version: %1").arg(info.new_version));
 
+    if (info.new_verison_type.has_value()) {
```
Comment (DioEgizio) [inline launcher/ui/dialogs/ModUpdateDialog.cpp]: ditto

File: launcher/ui/dialogs/ModUpdateDialog.cpp [hunk 12]
```diff
@@ -362,6 +362,11 @@ void ModUpdateDialog::appendMod(CheckUpdateTask::UpdatableMod const& info)
...
+        auto new_version_type_itme = new QTreeWidgetItem(item_top);
+        new_version_type_itme->setText(0, tr("New Version Type: %1").arg(info.new_verison_type.value().toString()));
```
Comment (DioEgizio) [inline launcher/ui/dialogs/ModUpdateDialog.cpp]: ditto

File: launcher/ui/pages/modplatform/ModPage.cpp [hunk 13]
```diff
@@ -137,8 +137,10 @@ void ModPage::updateVersionList()
         }
 
         // Only add the version if it's valid or using the 'Any' filter, but never if the version is opted out
-        if ((valid || m_filter->versions.empty()) && !optedOut(version))
-            m_ui->versionSelectionBox->addItem(version.version, QVariant(i));
+        if ((valid || m_filter->versions.empty()) && !optedOut(version)) {
+            auto release_type = version.verison_type.isValid() ? QString(" : %1").arg(version.verison_type.toString()) : "";
```
Comment (DioEgizio) [inline launcher/ui/pages/modplatform/ModPage.cpp]: ditto

File: launcher/ui/pages/modplatform/ResourcePage.cpp [hunk 14]
```diff
@@ -256,8 +256,8 @@ void ResourcePage::updateVersionList()
         auto& version = current_pack.versions[i];
         if (optedOut(version))
             continue;
-
-        m_ui->versionSelectionBox->addItem(current_pack.versions[i].version, QVariant(i));
+        auto release_type = current_pack.versions[i].verison_type.isValid() ? QString(" : %1").arg(current_pack.versions[i].verison_type.toString()) : "";
```
Comment (DioEgizio) [inline launcher/ui/pages/modplatform/ResourcePage.cpp]: ditto

**Comment by TheKodeToad** [comment]
lol you could've said that once :ioa:

**Comment by DioEgizio** [comment]
Better to be sure :ioa:

**Comment by Trial97** [comment]
this should fix #1222, right?

**Comment by Ryex** [comment]
> this should fix #1222, right?

not on it's own, no

**Comment by aoqia194** [comment]
Would love to see this get ready for merging! I was wondering why my select mods were updating to older release, turns out the beta is not accounted for. Was going to add this myself, but I am glad I searched this repository beforehand!

Obviously this does not fix #1222 which is ideal for me (like @Ryex said previous), but at least this gives functionality for the feature in the near future.

File: launcher/modplatform/ModIndex.cpp [hunk 15]
```diff
@@ -24,6 +24,74 @@
...
+}
+
+IndexedVersionType::Enum IndexedVersionType::enumFromString(const QString& type)
+{
+    auto found = s_indexed_version_type_names.constFind(type);
+    if (found != s_indexed_version_type_names.constEnd()) {
+        return *found;
+    } else {
+        return IndexedVersionType::Enum::UNKNOWN;
+    }
```
Comment (TheKodeToad) [inline launcher/modplatform/ModIndex.cpp]:
```suggestion
    return s_indexed_version_type_names.value(type, IndexedVersionType::Enum::UNKNOWN);
```

File: launcher/modplatform/ModIndex.cpp [hunk 16]
```diff
@@ -24,6 +24,74 @@
...
+            break;
+        case 2:
+            m_type = IndexedVersionType::Enum::Beta;
+            break;
+        case 3:
+            m_type = IndexedVersionType::Enum::Alpha;
+            break;
+        default:
+            m_type = IndexedVersionType::Enum::UNKNOWN;
+    }
```
Comment (TheKodeToad) [inline launcher/modplatform/ModIndex.cpp]: we shouldn't have cf-specific logic here imo
I guess the strings are ok, but the cf stuff is likely very specific to the platform

**Review by Trial97 — APPROVED** [review]

**Review by TheKodeToad — APPROVED** [review]

File: launcher/modplatform/ModIndex.cpp (hunk 16 above)
Comment (Trial97) [reply launcher/modplatform/ModIndex.cpp]: I think is better to be localized here instead of having it in a separate file.

Comment (TheKodeToad) [reply launcher/modplatform/ModIndex.cpp]: there's nothing in the function name to say it's curseforge-specific (well, it's a constructor lol)

Comment (Trial97) [reply launcher/modplatform/ModIndex.cpp]: Would renaming the variable `type` to `flame_type` help to clarify this?

Comment (TheKodeToad) [reply launcher/modplatform/ModIndex.cpp]: having it in this file is a bit of code smell in my opinion

Comment (Scrumplex) [reply launcher/modplatform/ModIndex.cpp]: The `type` param should just be the enum type

File: launcher/modplatform/ModIndex.h [hunk 17]
```diff
@@ -55,6 +56,35 @@ struct DonationData {
     QString url;
 };
 
+struct IndexedVersionType {
+    enum class Enum { Release = 1, Beta, Alpha, UNKNOWN };
```
Comment (Scrumplex) [inline launcher/modplatform/ModIndex.h]: why not something *like* this?
```suggestion
    enum class VersionType { Unknown = 0, Release, Beta, Alpha };
```

File: launcher/modplatform/ModIndex.cpp (hunk 16 above)
Comment (Trial97) [reply launcher/modplatform/ModIndex.cpp]: and do the logic(switch) in flame file or just remove this constructor and have a function here that transforms from flame type to our own enum?

File: launcher/modplatform/ModIndex.h (hunk 17 above)
Comment (Trial97) [reply launcher/modplatform/ModIndex.h]: Renamed the enum as suggested, but kept the order as is.
When sorting the release should be first(if they share the same date) not unknown.

File: launcher/modplatform/ModIndex.cpp (hunk 16 above)
Comment (Trial97) [reply launcher/modplatform/ModIndex.cpp]: After looking at the code more, I come to want to keep this constructor here, or at least if the constructor is too vague a function in ModIndex files that do this.
Let's wait for @Ryex opinion on this,

**Comment by ColonelGerdauf** [comment]
There is one point of awkwardness: curseforge

Due to what I assume to be difficulties in distinguishing between loader versions and incompatible sub-versions (1.19.x), mods are sometimes separated by loader into release/beta as a best alternative to having a proper tagging.

File: launcher/modplatform/ModIndex.cpp (hunk 16 above)
Comment (Trial97) [reply launcher/modplatform/ModIndex.cpp]: Removed that constructor

**Pull request merged**